src/Integration/Rules/SonarQubeQualityProfileProvider.cs

```diff
@@ -0,0 +1,135 @@
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            this.sonarQubeService = sonarQubeService;
+            this.logger = logger;
+        }
+
+        public QualityProfile GetQualityProfile(BoundSonarQubeProject project, Language language)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            if (language == null)
+            {
+                throw new ArgumentNullException(nameof(language));
+            }
+
+            // For now only C# and VB.Net have support for the connected mode in SLVS
+            if (language != Language.CSharp &&
+                language != Language.VBNET)
+            {
+                return null;
+            }
+
+            var roslynProfileExporter = GetSonarQubeQualityProfile(project, language).GetAwaiter().GetResult();
+            if (roslynProfileExporter == null)
+            {
+                return null;
+            }
+
+            var ruleset = ConvertToCodeAnalysisRuleSet(roslynProfileExporter);
+
+            return ConvertToSonarLintQualityProfile(ruleset, language);
+        }
+
+        private async Task<RoslynExportProfileResponse> GetSonarQubeQualityProfile(BoundSonarQubeProject project, Language language)
+        {
+            var serverLanguage = language.ToServerLanguage();
+
+            var qualityProfileInfo = await WebServiceHelper.SafeServiceCall(
+                () => this.sonarQubeService.GetQualityProfileAsync(project.ProjectKey, project.Organization?.Key, serverLanguage,
+                    CancellationToken.None),
+                this.logger);
+            if (qualityProfileInfo == null)
+            {
+                this.logger.WriteLine(string.Format(Strings.SubTextPaddingFormat,
+                   string.Format(Strings.CannotDownloadQualityProfileForLanguage, language.Name)));
+                return null;
+            }
+
+            var roslynProfileExporter = await WebServiceHelper.SafeServiceCall(
+                () => this.sonarQubeService.GetRoslynExportProfileAsync(qualityProfileInfo.Name, project.Organization?.Key,
+                    serverLanguage, CancellationToken.None),
+                this.logger);
+            if (roslynProfileExporter == null)
+            {
+                this.logger.WriteLine(string.Format(Strings.SubTextPaddingFormat,
+                    string.Format(Strings.QualityProfileDownloadFailedMessageFormat, qualityProfileInfo.Name,
+                        qualityProfileInfo.Key, language.Name)));
+                return null;
+            }
+
+            return roslynProfileExporter;
+        }
+
+        private RuleSet ConvertToCodeAnalysisRuleSet(RoslynExportProfileResponse roslynProfileExporter)
+        {
+            var tempRuleSetFilePath = Path.GetTempFileName();
+            File.WriteAllText(tempRuleSetFilePath, roslynProfileExporter.Configuration.RuleSet.OuterXml);
```

Comment:
File ops can fail unexpectedly e.g. if a virus checker is checking the file while we are trying to delete it.
It would be worth doing the write-read-delete in a try-catch, squashing non-critical exceptions.

Reply:
Ok but then is it wort re-trying (we have a retry helper)?

Reply:
I wouldn't bother with the retry. The write and the read should succeed since the user should have permission to their own temp folder. It's normally the delete that fails due to file locking, and it's not  a big problem if we leave the temp file behind; we just don't want to crash/fail to configure the analyzers correctly.